### PR TITLE
moreno20241120

### PR DESCRIPTION
**Pull request by MorenoMarrara**

### CODE REVIEW

File: client/pom.xml [hunk 1]
```diff
@@ -9,8 +9,8 @@
     <version>0.0.1</version>
 
     <properties>
-        <maven.compiler.source>21</maven.compiler.source>
-        <maven.compiler.target>21</maven.compiler.target>
+        <maven.compiler.source>17</maven.compiler.source>
```
Comment (albiosz) [inline client/pom.xml]: Why did you change the version of java?

File: shared/src/main/java/alcatraz/shared/Lobby.java [hunk 2]
```diff
@@ -6,42 +6,51 @@
...
         return this.players;
     }
 
     public String getOwner() {
         return this.ownerId;
     }
 
-    public boolean addPlayer(Player player) {
-        if (this.players.size() < 4) {
+    public void addPlayer(Player player) throws RemoteException {
```
Comment (albiosz) [inline shared/src/main/java/alcatraz/shared/Lobby.java]: The error could be more specific, the RemoteException is thrown when something is wrong with RMI connection, so it can be even misleading.
I think in this case we should create a new error e.g. LobbyIsFull or something like that.

File: shared/src/main/java/alcatraz/shared/LobbyKey.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+
+package alcatraz.shared;
+
+public class LobbyKey {
+    public final long id;
```
Comment (albiosz) [inline shared/src/main/java/alcatraz/shared/LobbyKey.java]: What for is the id of the key? Should not the secret already be unique, thus identifies the key 🤔

File: shared/pom.xml [hunk 4]
```diff
@@ -9,8 +9,8 @@
     <version>0.0.1</version>
 
     <properties>
-        <maven.compiler.source>21</maven.compiler.source>
-        <maven.compiler.target>21</maven.compiler.target>
+        <maven.compiler.source>17</maven.compiler.source>
```
Comment (albiosz) [inline shared/pom.xml]: why?

File: server/pom.xml [hunk 5]
```diff
@@ -9,8 +9,8 @@
     <version>0.0.1</version>
 
     <properties>
-        <maven.compiler.source>21</maven.compiler.source>
-        <maven.compiler.target>21</maven.compiler.target>
+        <maven.compiler.source>17</maven.compiler.source>
```
Comment (albiosz) [inline server/pom.xml]: why?

File: server/src/main/java/alcatraz/server/state/LobbyManager.java [hunk 6]
```diff
@@ -1,34 +1,78 @@
...
 
-    public void createLobby(long lobbyId,Lobby lobby) {
-        lobbies.put(lobbyId, lobby);
+    /*
+    * @brief: Construct a new lobby and assign it to the creator's name.
+    *           Several lobbies can be created by the same owner but only the creator will get the secret.
+    *           When an owner creates a new lobby and they have been inside a lobby before, they will be relocated to
+    *               the newly created lobby.
+     */
+    public LobbyKey createLobby(Player owner) throws RemoteException {
```
Comment (albiosz) [inline server/src/main/java/alcatraz/server/state/LobbyManager.java:31]: the exception should be more specific

File: server/src/main/java/alcatraz/server/state/LobbyManager.java [hunk 7]
```diff
@@ -1,34 +1,78 @@
...
     }
+
     public Lobby getLobbyById(long lobbyId){
         return lobbies.get(lobbyId);
     }
-    public Boolean addPlayerToLobby(long lobbyId, Player player){
-        // TODO: it needs to be checked of there is still space for another player, maximal 4 players
-        return lobbies.get(lobbyId).addPlayer(player); // vielleicht kann das ein void sein mit exception?
+
+    public void addPlayerToLobby(long lobbyId, Player player) throws RemoteException {
```
Comment (albiosz) [inline server/src/main/java/alcatraz/server/state/LobbyManager.java:81]: the exception could be more precise

File: server/src/main/java/alcatraz/server/state/LobbyManager.java [hunk 8]
```diff
@@ -1,34 +1,78 @@
...
+        lobbies.get(lobbyId).addPlayer(player);
+    }
+
+    public void removePlayerFromLobby(String playerName) throws RemoteException {
```
Comment (albiosz) [inline server/src/main/java/alcatraz/server/state/LobbyManager.java]: the exception could be more precise

File: server/src/main/java/alcatraz/server/state/LobbyManager.java [hunk 9]
```diff
@@ -1,34 +1,78 @@
...
+    *           Several lobbies can be created by the same owner but only the creator will get the secret.
+    *           When an owner creates a new lobby and they have been inside a lobby before, they will be relocated to
+    *               the newly created lobby.
+     */
+    public LobbyKey createLobby(Player owner) throws RemoteException {
+        if (lobbies.size() == MAXSIZE) {
+            throw new RemoteException();        //MM20241122: find better Exception! "MAXSIZE reached"
+        }
+
+        while (lobbies.containsKey(lobbyIdCounter)) {
```
Comment (albiosz) [inline server/src/main/java/alcatraz/server/state/LobbyManager.java:36]: Is it really more efficient to look for a first free id, instead of generating random lobby id? 🤔

File: server/src/main/java/alcatraz/server/Server.java [hunk 10]
```diff
@@ -22,68 +19,69 @@ public Server(ReplicationInterface replication) throws RemoteException {
...
             // Store the Player with ClientInterface
-            this.state.players.put(playerName, new Player(client, playerName, "ip_address", "port"));
-            System.out.println(playerName + " registered");
-            return true;
+            //MM20241121: use a mutex at this stage!
+            this.state.players.put(clientName, new Player(client, clientName, "ip_address", "port"));
+            System.out.println(clientName + " registered");
+            this.replication.replicatePrimaryState();
             // TODO: there should be an early return instead of else
+            //MM20241121: why?
```
Comment (albiosz) [inline server/src/main/java/alcatraz/server/Server.java]: It would demand less cognitive effort.
```java
if (!isNotPrimary) {
  forward();
  return;
}
// the rest function logic
```
Now we have to find the else every time that is 13 lines under the `if`, and then have to look even further if it is the last statement in the function. With early return you can see directly all the actions that happen if the server is not primary.

File: server/src/main/java/alcatraz/server/Server.java [hunk 11]
```diff
@@ -22,68 +19,69 @@ public Server(ReplicationInterface replication) throws RemoteException {
...
-            Lobby lobby = new Lobby(lobbyId, lobbyPlayers, clientName);
-            this.state.lobbyManager.createLobby(lobbyId, lobby);
-            System.out.println("Lobby created with ID: " + lobbyId + " by Player: " + clientName);
-
-            // Replication logic: Update backups with the new lobbies state
+            //MM20241121: This section needs a mutex
+            LobbyKey key = this.state.lobbyManager.createLobby(owner);
+            System.out.println("Lobby created with ID: " + key.id + " by Player: " + owner.getClientName());
+
+            this.state.lobbyManager.addPlayerToLobby(key.id, owner);
```
Comment (albiosz) [inline server/src/main/java/alcatraz/server/Server.java]: we also will need a mutex here, won't we? 🤔

File: server/src/main/java/alcatraz/server/Server.java [hunk 12]
```diff
@@ -22,68 +19,69 @@ public Server(ReplicationInterface replication) throws RemoteException {
...
+            //              -> Therefore the client must check if creation was successful when an exception was thrown!
+            return this.replication.getPrimaryServer().createLobby(owner);
         }
     }
 
     // TODO: refactor nested if statements
     @Override
-    public boolean joinLobby(String clientName, Long lobbyId) throws RemoteException {
+    public void joinLobby(Player client, Long lobbyId) throws RemoteException {
         if (this.replication.isPrimary()) {
```
Comment (albiosz) [inline server/src/main/java/alcatraz/server/Server.java]: I would decrease the cyclomatic complexity of the function. From my perspective it is no not really readable now.

I think something like this would be more readable and will grow better in case there is more logic to add.
```java
if (!isPrimary) {
  // forwarding
  return
}
if (!doesNotContainKey) {
  // the exception
  return
}
// the rest
```

File: client/src/main/java/alcatraz/client/Main.java [hunk 13]
```diff
@@ -69,8 +70,8 @@ public static void main(String[] args) {
                         break;
                     case "2":
                         System.out.println("Creating Lobby...");
-                        Lobby lobby = server.createLobby(clientName);
-                        System.out.println("Lobby created with id: " + lobby.getId());
+                        LockedLobby lobby = server.createLobby(clientName);
```
Comment (albiosz) [inline client/src/main/java/alcatraz/client/Main.java]: I got an error a compilation error because the `LockeLobby` was not found

**Review by albiosz — CHANGES_REQUESTED** [review]
The code does not compile, please correct the errors

File: shared/src/main/java/alcatraz/shared/DuplicateNameException.java [hunk 14]
```diff
@@ -0,0 +1,10 @@
+package alcatraz.shared;
```
Comment (albiosz) [inline shared/src/main/java/alcatraz/shared/DuplicateNameException.java]: the exception could me moved to a separate sub package "exceptions", to not mix with other classes that are responsible for the game logic

File: shared/src/main/java/alcatraz/shared/LobbyFullException.java [hunk 15]
```diff
@@ -0,0 +1,9 @@
+package alcatraz.shared;
```
Comment (albiosz) [inline shared/src/main/java/alcatraz/shared/LobbyFullException.java]: the exception could me moved to a separate sub package "exceptions", to not mix with other classes that are responsible for the game logic

File: shared/src/main/java/alcatraz/shared/LobbyLockedException.java [hunk 16]
```diff
@@ -0,0 +1,9 @@
+package alcatraz.shared;
```
Comment (albiosz) [inline shared/src/main/java/alcatraz/shared/LobbyLockedException.java]: move the exception to the separate sub package

File: shared/src/main/java/alcatraz/shared/NotEnoughPlayersException.java [hunk 17]
```diff
@@ -0,0 +1,11 @@
+package alcatraz.shared;
```
Comment (albiosz) [inline shared/src/main/java/alcatraz/shared/NotEnoughPlayersException.java]: the exception could me moved to a separate sub package "exceptions", to not mix with other classes that are responsible for the game logic

File: shared/src/main/java/alcatraz/shared/PlayerNotRegisteredException.java [hunk 18]
```diff
@@ -0,0 +1,10 @@
+package alcatraz.shared;
```
Comment (albiosz) [inline shared/src/main/java/alcatraz/shared/PlayerNotRegisteredException.java]: the exception could me moved to a separate sub package "exceptions", to not mix with other classes that are responsible for the game logic

File: shared/src/main/java/alcatraz/shared/TooManyLobbiesException.java [hunk 19]
```diff
@@ -0,0 +1,10 @@
+package alcatraz.shared;
```
Comment (albiosz) [inline shared/src/main/java/alcatraz/shared/TooManyLobbiesException.java]: the exception could me moved to a separate sub package "exceptions", to not mix with other classes that are responsible for the game logic

File: server/src/main/java/alcatraz/server/Server.java [hunk 20]
```diff
@@ -2,146 +2,150 @@
...
+                fairLock.unlock();
+                throw new DuplicateNameException(clientName);
             }
+
             // Store the Player with ClientInterface
-            this.state.players.put(playerName, new Player(client, playerName, "ip_address", "port"));
-            System.out.println(playerName + " registered");
-            return true;
-            // TODO: there should be an early return instead of else
+            //MM20241121: use a mutex at this stage!
```
Comment (albiosz) [inline server/src/main/java/alcatraz/server/Server.java]: outdated comment

File: server/src/main/java/alcatraz/server/Server.java [hunk 21]
```diff
@@ -2,146 +2,150 @@
...
+            this.state.players.put(clientName, new Player(client, clientName, "ip_address", "port"));
```
Comment (albiosz) [inline server/src/main/java/alcatraz/server/Server.java]: ip_address and port should come from the client.
It would be nice to add TODO at this point and implement the logic in another PR,
or implement it directly, by changing the signature of the function to include `ipAddress` and `port` too

File: server/src/main/java/alcatraz/server/Server.java [hunk 22]
```diff
@@ -2,146 +2,150 @@
...
         if (this.replication.isPrimary()) {
-            SharedState.lobbyIdCounter++;
-            long lobbyId = SharedState.lobbyIdCounter;
-            Map<String, Player> lobbyPlayers = new HashMap<>();
-            lobbyPlayers.put(clientName, this.state.players.get(clientName));
-            Lobby lobby = new Lobby(lobbyId, lobbyPlayers, clientName);
-            this.state.lobbyManager.createLobby(lobbyId, lobby);
-            System.out.println("Lobby created with ID: " + lobbyId + " by Player: " + clientName);
+            fairLock.lock();
+            //MM20241121: This section needs a mutex
```
Comment (albiosz) [inline server/src/main/java/alcatraz/server/Server.java]: outdated comment

File: server/src/main/java/alcatraz/server/Server.java [hunk 23]
```diff
@@ -2,146 +2,150 @@
...
-    public void initializeGameStart(long lobbyId) throws RemoteException {
+    public void initializeGameStart(long lobbyId, String secret) throws RemoteException {
         if (this.replication.isPrimary()) {
-            Lobby lobby = this.state.lobbyManager.getLobbyById(lobbyId);
-            lobby.getPlayers().forEach((key, value) -> {
+            this.fairLock.lock();
+            Lobby gameLobby = this.state.lobbyManager.getLobbyById(lobbyId);
+            if (gameLobby.checkSecret(secret)) {
+                gameLobby.setUnavailable();
+                gameLobby.getPlayers().forEach((key, value) -> {
```
Comment (albiosz) [inline server/src/main/java/alcatraz/server/Server.java]: change the names of the parameters (key, value) to the names that describe what the variables contain e.g. (playerName, player)

**Review by albiosz — CHANGES_REQUESTED** [review]
- I would refactor all the methods in the Server.java class, so that they early return where it is possible, instead of `else` statements
e.g.
```java
if (!primary) {
  replicate();
  return;
}

if (!checkSecret()) {
    throw new CheckSecretException();
}
```
It should considerable improve the quality (lower cyclomatic complexity) of the code.

File: shared/src/main/java/alcatraz/shared/Lobby.java (hunk 2 above)
Comment (MorenoMarrara) [reply shared/src/main/java/alcatraz/shared/Lobby.java]: see LobbyFullException

File: shared/src/main/java/alcatraz/shared/LobbyKey.java (hunk 3 above)
Comment (MorenoMarrara) [reply shared/src/main/java/alcatraz/shared/LobbyKey.java]: was renamed to lobbyId to make it clearer

File: server/src/main/java/alcatraz/server/state/LobbyManager.java (hunk 6 above)
Comment (MorenoMarrara) [reply server/src/main/java/alcatraz/server/state/LobbyManager.java:31]: "throws RemoteException" was removed as method it shouldn't throw anyways

File: server/src/main/java/alcatraz/server/state/LobbyManager.java (hunk 9 above)
Comment (MorenoMarrara) [reply server/src/main/java/alcatraz/server/state/LobbyManager.java:36]: it was the simplest mechanism, that I could think of and that also works

File: server/src/main/java/alcatraz/server/state/LobbyManager.java (hunk 7 above)
Comment (MorenoMarrara) [reply server/src/main/java/alcatraz/server/state/LobbyManager.java:81]: I guess it can't, as removePlayerFromLobby() throws LockedLobbyException and addPlayer() throws RemoteException

File: server/src/main/java/alcatraz/server/Server.java (hunk 10 above)
Comment (MorenoMarrara) [reply server/src/main/java/alcatraz/server/Server.java]: totally agree! didn't know what you meant by early exit!

File: server/src/main/java/alcatraz/server/Server.java (hunk 21 above)
Comment (MorenoMarrara) [reply server/src/main/java/alcatraz/server/Server.java]: I agree, but does the server have to authenticate the ip_address or does it simply believe what it's told?

File: shared/src/main/java/alcatraz/shared/DuplicateNameException.java (hunk 14 above)
Comment (MorenoMarrara) [reply shared/src/main/java/alcatraz/shared/DuplicateNameException.java]: totally agree; I suggest, we make a sub package "alcatraz.shared.Interfaces", "alcatraz.shared.Exceptions", "alcattraz.shared.Utils"

File: shared/src/main/java/alcatraz/shared/LobbyFullException.java (hunk 15 above)
Comment (MorenoMarrara) [reply shared/src/main/java/alcatraz/shared/LobbyFullException.java]: totally agree; I suggest, we make a sub package "alcatraz.shared.Interfaces", "alcatraz.shared.Exceptions", "alcattraz.shared.Utils"

File: shared/src/main/java/alcatraz/shared/LobbyLockedException.java (hunk 16 above)
Comment (MorenoMarrara) [reply shared/src/main/java/alcatraz/shared/LobbyLockedException.java]: totally agree; I suggest, we make a sub package "alcatraz.shared.Interfaces", "alcatraz.shared.Exceptions", "alcattraz.shared.Utils"

File: shared/src/main/java/alcatraz/shared/NotEnoughPlayersException.java (hunk 17 above)
Comment (MorenoMarrara) [reply shared/src/main/java/alcatraz/shared/NotEnoughPlayersException.java]: totally agree; I suggest, we make a sub package "alcatraz.shared.Interfaces", "alcatraz.shared.Exceptions", "alcattraz.shared.Utils"

File: shared/src/main/java/alcatraz/shared/PlayerNotRegisteredException.java (hunk 18 above)
Comment (MorenoMarrara) [reply shared/src/main/java/alcatraz/shared/PlayerNotRegisteredException.java]: totally agree; I suggest, we make a sub package "alcatraz.shared.Interfaces", "alcatraz.shared.Exceptions", "alcattraz.shared.Utils"

File: shared/src/main/java/alcatraz/shared/TooManyLobbiesException.java (hunk 19 above)
Comment (MorenoMarrara) [reply shared/src/main/java/alcatraz/shared/TooManyLobbiesException.java]: totally agree; I suggest, we make a sub package "alcatraz.shared.Interfaces", "alcatraz.shared.Exceptions", "alcattraz.shared.Utils"

File: server/src/main/java/alcatraz/server/Server.java (hunk 21 above)
Comment (albiosz) [reply server/src/main/java/alcatraz/server/Server.java]: I'd say that in the first version the server should trust, that the client passed a correct IP and in further versions we can improve the process, if there is need (like using the function `client.isAvailable()`)

Comment (albiosz) [reply server/src/main/java/alcatraz/server/Server.java]: I changed the this thread to a separate issue

**Review by albiosz — APPROVED** [review]
